File: server/routes/apiKeys.ts
/**
 * API key management (session-cookie auth) — issuing/listing/revoking the
 * bearer credentials POST /api/v1/predict/:runId accepts (see
 * routes/api-v1.ts and the `apiKeyAuth` macro in routes/auth.ts).
 *
 * - POST   /api/keys           → Create a key, returns the raw key ONCE — never stored, never shown again
 * - GET    /api/keys           → List the caller's keys (prefix + metadata only, never the raw key)
 * - DELETE /api/keys/:keyId    → Revoke a key (soft — keeps the audit trail of what it was)
 */

import { db } from '@server/db'
import { apiKeys } from '@server/db/schema'
import { generateApiKey } from '@server/lib/apiKeys'
import { and, eq, isNull } from 'drizzle-orm'
import { Elysia, status, t } from 'elysia'
import { betterAuth } from './auth'

export const apiKeyRoutes = new Elysia({ prefix: '/api/keys' })
  .use(betterAuth)
  /* ── Create a new API key ── */
  .post(
    '/',
    async ({ user, body }) => {
      const { rawKey, keyHash, keyPrefix } = generateApiKey()
      const [key] = await db
        .insert(apiKeys)
        .values({ userId: user.id, name: body.name, keyHash, keyPrefix })
        .returning({ id: apiKeys.id, name: apiKeys.name, keyPrefix: apiKeys.keyPrefix, createdAt: apiKeys.createdAt })

      // The only point in this key's lifetime the raw value is ever
      // available — it is not derivable from keyHash, so losing this
      // response means generating a new key.
      return status(201, { ...key, key: rawKey })
    },
    {
      auth: true,
      body: t.Object({ name: t.String({ minLength: 1, maxLength: 100 }) }),
    },
  )
  /* ── List the caller's keys — never the raw value, only what was shown at creation ── */
  .get(
    '/',
    async ({ user }) => {
      const keys = await db.query.apiKeys.findMany({
        where: { userId: user.id },
        columns: { id: true, name: true, keyPrefix: true, lastUsedAt: true, createdAt: true, revokedAt: true },
        orderBy: { createdAt: 'desc' },
      })
      return { keys }
    },
    { auth: true },
  )
  /* ── Revoke a key ── */
  .delete(
    '/:keyId',
    async ({ user, params }) => {
      const [revoked] = await db
        .update(apiKeys)
        .set({ revokedAt: new Date() })
        .where(and(eq(apiKeys.id, params.keyId), eq(apiKeys.userId, user.id), isNull(apiKeys.revokedAt)))
        .returning({ id: apiKeys.id })
      if (!revoked) return status(404, 'API key not found')
      return status(204)
    },
    {
      auth: true,
      params: t.Object({ keyId: t.String({ format: 'uuid' }) }),
    },
  )

File: ai_service/theseus/routers/api_keys.py
"""/api/keys: issue, list and revoke the bearer keys the hosted prediction API (/api/v1) accepts."""

import uuid

import sqlalchemy as sa
from fastapi import APIRouter, HTTPException

from theseus.auth.api_keys import generate_api_key
from theseus.db.models import ApiKey
from theseus.deps import SessionDep, UserId
from theseus.schemas.api_keys import ApiKeyListResponse, ApiKeyOut, CreateApiKeyBody, CreatedApiKey

router = APIRouter(prefix="/keys", tags=["api-keys"])


@router.post("", status_code=201, response_model=CreatedApiKey)
async def create_api_key(body: CreateApiKeyBody, user_id: UserId, session: SessionDep) -> CreatedApiKey:
    raw, key_hash, key_prefix = generate_api_key()
    key = ApiKey(user_id=user_id, name=body.name, key_hash=key_hash, key_prefix=key_prefix)
    session.add(key)
    await session.commit()
    await session.refresh(key)
    # The only moment the raw value exists: it is not derivable from the hash, so losing this
    # response means generating a new key.
    return CreatedApiKey(id=key.id, name=key.name, key_prefix=key.key_prefix, created_at=key.created_at, key=raw)


@router.get("", response_model=ApiKeyListResponse)
async def list_api_keys(user_id: UserId, session: SessionDep) -> ApiKeyListResponse:
    keys = (
        (await session.execute(sa.select(ApiKey).where(ApiKey.user_id == user_id).order_by(ApiKey.created_at.desc())))
        .scalars()
        .all()
    )
    return ApiKeyListResponse(keys=[ApiKeyOut.model_validate(k) for k in keys])


@router.delete("/{key_id}", status_code=204)
async def revoke_api_key(key_id: uuid.UUID, user_id: UserId, session: SessionDep) -> None:
    """Soft revoke: keeps the audit trail of what a (possibly leaked) key was."""
    res = await session.execute(
        sa.update(ApiKey)
        .where(ApiKey.id == key_id, ApiKey.user_id == user_id, ApiKey.revoked_at.is_(None))
        .values(revoked_at=sa.func.now())
        .returning(ApiKey.id)
    )
    if res.first() is None:
        raise HTTPException(404, "API key not found")
    await session.commit()
